File: src/main/java/uk/edu/populationfitness/models/genes/performance/GenesTimerFactory.java
package uk.edu.populationfitness.models.genes.performance;

import uk.edu.populationfitness.models.Config;
import uk.edu.populationfitness.models.genes.Function;
import uk.edu.populationfitness.models.genes.Genes;
import uk.edu.populationfitness.models.genes.GenesFactory;

public class GenesTimerFactory implements GenesFactory {
    private final GenesFactory delegate;

    public GenesTimerFactory(GenesFactory delegate) {
        this.delegate = delegate;
    }

    @Override
    public Genes build(Config config) {
        Genes genes = delegate.build(config);
        return new GenesTimer(genes);
    }

    @Override
    public void useFitnessFunction(Function function) {
        delegate.useFitnessFunction(function);
    }

    @Override
    public Function getFitnessFunction() {
        return delegate.getFitnessFunction();
    }
}

File: DotNet/PopulationFitness/PopulationFitness/Models/Genes/Performance/GenesTimerFactory.cs
namespace PopulationFitness.Models.Genes.Performance
{
    public class GenesTimerFactory : IGenesFactory
    {
        private readonly IGenesFactory _factory;

        public GenesTimerFactory(IGenesFactory factory) => this._factory = factory;

        public IGenes Build(Config config) => new GenesTimer(_factory.Build(config));

        public Function FitnessFunction { get { return _factory.FitnessFunction; } set { _factory.FitnessFunction = value; } }
    }
}
